### PR TITLE
SdkRemoteLaunchPlans

### PR DESCRIPTION
**Pull request by ckiosidis**
# TL;DR

Adding support for SdkRemoteLaunchPlan, similarly to the existing SdkRemoteTask.
I will put together a separate PR for the SdkRemoteWorkflow 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Created a SdkRemoteLaunchPlan class that implements the apply method to return a SdkNode.
The SdkWorkflowNode/WorkflowNode classes were already implemented previously.

## Tracking Issue
adds to  https://github.com/flyteorg/flyte/issues/1002

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  @Nullable
+  public abstract String domain();
+
+  public abstract String project();
+
+  public abstract String name();
+
+  @Nullable
+  public String version() {
```
Comment (tnsetting) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:40]: How would version work?

Comment (ckiosidis) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:40]: When the version is empty, the sdk fetches it from the platform, see https://github.com/flyteorg/flytekit-java/blob/master/jflyte/src/main/java/org/flyte/jflyte/IdentifierRewrite.java#L140-L161

Comment (tnsetting) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:40]: Can we add a test for it somehow?

Comment (ckiosidis) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:40]: The IdentifierRewrite is tested separately, and I added the example here that works when I tested it https://github.com/flyteorg/flytekit-java/blob/b8b5e0e677be7da6d123b81c511e9283790b1484/flytekit-examples/src/main/java/org/flyte/examples/RemoteLaunchPlanExample.java

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkRemoteLaunchPlanTest.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                        .id("some-node-id")
+                        .workflowNode(
+                            WorkflowNode.builder()
+                                .reference(
+                                    WorkflowNode.Reference.ofLaunchPlanRef(
+                                        PartialLaunchPlanIdentifier.builder()
+                                            .domain("dev")
+                                            .project("project-a")
+                                            .name("SomeWorkflow")
+                                            .version("version")
```
Comment (tnsetting) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkRemoteLaunchPlanTest.java:72]: should be the `version` be null?

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            .domain(domain())
+            .version(version())
+            .build();
+    List<CompilerError> errors = Compiler.validateApply(nodeId, inputs, inputs().getVariableMap());
+
+    if (!errors.isEmpty()) {
+      throw new CompilerException(errors);
+    }
+
+    Map<String, SdkBindingData> outputs =
```
Comment (tnsetting) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:83]: So we don't validate the input/output of the remote Launch plan.

Comment (ckiosidis) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:83]: We don't double check the remote outputs but the register step still fails if the interface is difference.
This part I copied from another place to keep it consistent

Comment (tnsetting) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java:83]: Ok. Cool.

**Review by tnsetting — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]
With my limited knowledge in this SDK, it looks good to me!
